File: .NET Framework 4 Base64 Converter/[C#][VB]-.NET Framework 4 Base64 Converter/C# and VB/C#,VB/Base64ConverterCS/Encode64.cs
using System;
using System.IO;
using System.Reflection;
using System.Resources;
using System.Security;
using System.Security.Cryptography;
using System.Text;

class Encode64
{
   public const int DefaultSize = 1024000;

   private static int BufferSize = DefaultSize;
   private static int CharsToRead = DefaultSize / 2;
   private static ResourceManager rm;

   public static void Main()
   {
      string[] args = Environment.GetCommandLineArgs();
      rm = new ResourceManager("Example.CmdResources", Assembly.GetExecutingAssembly());
      bool toEncode = false;
      bool toDecode = false;
      
      // Display application name.
      Console.WriteLine(rm.GetString("AppName"));
            
      // Display help text if command line includes /?
      if (args.Length < 3 | Array.Exists(args, new Switches("/?").IsSwitchSet))
      {
         Console.WriteLine(rm.GetString("HelpText"));
         return;
      }

      if (Array.Exists(args, new Switches("/e").IsSwitchSet)) 
         toEncode = true;
      if (Array.Exists(args, new Switches("/d").IsSwitchSet)) 
         toDecode = true;
      if (Array.Exists(args, new Switches("/s").HasSwitch))
      {
         BufferSize = Switches.GetBufferSize(args);
         CharsToRead = BufferSize / 2;
      }

      // Make sure only an encode or a decode switch is present.
      if (toEncode & toDecode) {
         Console.WriteLine(rm.GetString("BadSwitches"));
         return;
      }
      else if (! toEncode & ! toDecode) {
         toEncode = true;
      }

      // Make sure the source file exists.
      string srcFile = args[args.Length - 2];
      string targetFile = args[args.Length - 1];
      if (! File.Exists(srcFile)) {
         Console.WriteLine(rm.GetString("FileNotFound"), srcFile);
         return;
      }

      // Perform the encoding or the decoding.
      if (toEncode)
         EncodeFile(srcFile, targetFile);
      else
         DecodeFile(srcFile, targetFile);
   }

   private static void EncodeFile(string srcFile , string targetFile)
   {
      byte[] bytes = new byte[BufferSize];
      FileStream fs = null;
      StreamWriter sw = null;
      FileStream fsOut = null;
      int bytesRead = 0;
      int totalBytesRead = 0;

      // Open srcFile in read-only mode.
      try {
         fs = new FileStream(srcFile, FileMode.Open, FileAccess.Read, FileShare.Read, BufferSize);
         long sourceSize = new FileInfo(srcFile).Length;

         if (sourceSize <= BufferSize) {
            // Open stream writer
            sw = new StreamWriter(targetFile, false, Encoding.ASCII);
            bytesRead = fs.Read(bytes, 0, BufferSize);

            if (bytesRead > 0) {
               string base64String = Convert.ToBase64String(bytes, 0, bytesRead);
               totalBytesRead += bytesRead;
               Console.CursorLeft = 0;
               Console.Write(rm.GetString("ProgressEncodedMsg"), totalBytesRead);
               sw.Write(base64String);
            }
         }
         else {
            // Instantiate a ToBase64Transform object.
            ToBase64Transform transf = new ToBase64Transform();
            // Arrays to hold input and output bytes.
            byte[] inputBytes = new byte[transf.InputBlockSize];
            byte[] outputBytes = new byte[transf.OutputBlockSize];
            int bytesWritten;

            fsOut = new FileStream(targetFile, FileMode.Create, FileAccess.Write);

            do {
               bytesRead = fs.Read(inputBytes, 0, inputBytes.Length);
               totalBytesRead += bytesRead;
               bytesWritten = transf.TransformBlock(inputBytes, 0, bytesRead, outputBytes, 0);
               fsOut.Write(outputBytes, 0, bytesWritten);
            } while (sourceSize - totalBytesRead > transf.InputBlockSize);

            // Transform the final block of data.
            bytesRead = fs.Read(inputBytes, 0, inputBytes.Length);
            byte[] finalOutputBytes = transf.TransformFinalBlock(inputBytes, 0, bytesRead);
            fsOut.Write(finalOutputBytes, 0, finalOutputBytes.Length);

            // Clear Base64Transform object.
            transf.Clear();
         }

         Console.WriteLine();
         Console.WriteLine(rm.GetString("SuccessEncodedMsg"), srcFile, targetFile);
      }
      catch (IOException) {
         Console.WriteLine(rm.GetString("IOException"));
         return;
      }
      catch (SecurityException) {
         Console.WriteLine(rm.GetString("SecurityException"), srcFile);
         return;
      }
      catch (UnauthorizedAccessException) {
         Console.WriteLine(rm.GetString("AccessException"), srcFile);
         return;
      }
      finally {
         if (sw != null) sw.Close();
         if (fs != null) {
            fs.Dispose();
            fs.Close();
         }
         if (fsOut != null) {
            fsOut.Dispose();
            fsOut.Close();
         }
      }
   }

   private static void DecodeFile(string srcFile , string targetFile)
   {

      TextReader reader = null;
      FileStream fs = null;
      FileStream fsIn = null;
      long sourceSize = new FileInfo(srcFile).Length;

      try {
         fs = new FileStream(targetFile, FileMode.Create, FileAccess.Write);

         // Read entire file in a single operation.
         if (sourceSize <= CharsToRead) {
            reader = new StreamReader(srcFile, Encoding.ASCII);

            string encodedChars = reader.ReadToEnd();

            try {
               byte[] bytes = Convert.FromBase64String(encodedChars);
               fs.Write(bytes, 0, bytes.Length);
            }
            catch (FormatException e) {
               Console.WriteLine(e.Message);
            }

            Console.CursorLeft = 0;
         }
         else {
            // Read file as a stream into a buffer.

            // Instantiate a FromBase64Transform object.
            FromBase64Transform  transf = new FromBase64Transform();
            // Arrays to hold input and output bytes.
            byte[] inputBytes = new byte[transf.InputBlockSize];
            byte[] outputBytes = new byte[transf.OutputBlockSize];
            int bytesRead = 0;
            int bytesWritten  = 0;
            int totalBytesRead = 0;

            fsIn = new FileStream(srcFile, FileMode.Open, FileAccess.Read);

            do {
               bytesRead = fsIn.Read(inputBytes, 0, inputBytes.Length);
               totalBytesRead += bytesRead;
               bytesWritten = transf.TransformBlock(inputBytes, 0, bytesRead, outputBytes, 0);
               fs.Write(outputBytes, 0, bytesWritten);
            } while (sourceSize - totalBytesRead > transf.InputBlockSize);

            // Transform the final block of data.
            bytesRead = fsIn.Read(inputBytes, 0, inputBytes.Length);
            byte[] finalOutputBytes = transf.TransformFinalBlock(inputBytes, 0, bytesRead);
            fs.Write(finalOutputBytes, 0, finalOutputBytes.Length);

            // Clear Base64Transform object.
            transf.Clear();
         }

         Console.WriteLine();
         Console.WriteLine(rm.GetString("SuccessDecodedMsg"), srcFile, targetFile);
      }
      catch (IOException) {
         Console.WriteLine(rm.GetString("IOException"));
         return;
      }
      finally {
         if (reader != null)
            reader.Close();
         if (fs != null) {
            fs.Close(); 
            fs.Dispose();
         }
         if (fsIn != null)
         {
            fsIn.Close();
            fs.Dispose();
         }
      }
   }
}

File: .NET Framework 4 Base64 Converter/[C#][VB]-.NET Framework 4 Base64 Converter/C# and VB/C#,VB/Base64ConverterVB/Encode64.vb
Imports System.IO
Imports System.Resources
Imports System.Security
Imports System.Security.Cryptography
Imports System.Text

Module Encode64

   Public Const DefaultSize As Integer = 1024000

   Private BufferSize As Integer = DefaultSize
   Private CharsToRead As Integer = BufferSize \ 2

   Sub Main()
      Dim args() As String = Environment.GetCommandLineArgs()
      Dim toEncode, toDecode As Boolean

      ' Display application name.
      Console.WriteLine(My.Resources.CmdResources.AppName)

      ' Display help text if command line includes /?
      If args.Length < 3 Or Array.Exists(args, AddressOf New Switches("/?").IsSwitchSet) Then
         Console.WriteLine(My.Resources.CmdResources.HelpText)
         Exit Sub
      End If

      If Array.Exists(args, AddressOf New Switches("/e").IsSwitchSet) Then toEncode = True
      If Array.Exists(args, AddressOf New Switches("/d").IsSwitchSet) Then toDecode = True
      If Array.Exists(args, AddressOf New Switches("/s").HasSwitch) Then
         BufferSize = Switches.GetBufferSize(args)
         CharsToRead = BufferSize \ 2
      End If

      ' Make sure only an encode or a decode switch is present.
      If toEncode And toDecode Then
         Console.WriteLine(My.Resources.CmdResources.BadSwitches)
         Exit Sub
      ElseIf Not toEncode And Not toDecode Then
         toEncode = True
      End If

      ' Make sure the source file exists.
      Dim srcFile As String = args(args.Length - 2)
      Dim targetFile As String = args(args.Length - 1)
      If Not File.Exists(srcFile) Then
         Console.WriteLine(My.Resources.CmdResources.FileNotFound, srcFile)
         Exit Sub
      End If

      ' Perform the encoding or the decoding.
      If toEncode Then
         EncodeFile(srcFile, targetFile)
      Else
         DecodeFile(srcFile, targetFile)
      End If
   End Sub

   Private Sub EncodeFile(ByVal srcFile As String, ByVal targetFile As String)
      Dim bytes(BufferSize - 1) As Byte
      Dim fs As FileStream = Nothing
      Dim sw As StreamWriter = Nothing
      Dim fsOut As FileStream = Nothing
      Dim bytesRead As Integer = 0
      Dim totalBytesRead As Integer = 0

      ' Open srcFile in read-only mode.
      Try
         fs = New FileStream(srcFile, FileMode.Open, FileAccess.Read, FileShare.Read, BufferSize)
         Dim sourceSize As Long = New FileInfo(srcFile).Length

         If sourceSize <= BufferSize Then
            ' Open stream writer
            sw = New StreamWriter(targetFile, False, Encoding.ASCII)
            bytesRead = fs.Read(bytes, 0, BufferSize)

            If bytesRead > 0 Then
               Dim base64String As String = Convert.ToBase64String(bytes, 0, bytesRead)
               totalBytesRead += bytesRead
               Console.CursorLeft = 0
               Console.Write(My.Resources.CmdResources.ProgressEncodedMsg, totalBytesRead)
               sw.Write(base64String)
            End If
         Else
            ' Instantiate a ToBase64Transform object.
            Dim transf As New ToBase64Transform()
            ' Arrays to hold input and output bytes.
            Dim inputBytes(transf.InputBlockSize - 1) As Byte
            Dim outputBytes(transf.OutputBlockSize - 1) As Byte
            Dim bytesWritten As Integer

            fsOut = New FileStream(targetFile, FileMode.Create, FileAccess.Write)

            Do
               bytesRead = fs.Read(inputBytes, 0, inputBytes.Length)
               totalBytesRead += bytesRead
               bytesWritten = transf.TransformBlock(inputBytes, 0, bytesRead, outputBytes, 0)
               fsOut.Write(outputBytes, 0, bytesWritten)
            Loop While sourceSize - totalBytesRead > transf.InputBlockSize

            ' Transform the final block of data.
            bytesRead = fs.Read(inputBytes, 0, inputBytes.Length)
            Dim finalOutputBytes As Byte() = transf.TransformFinalBlock(inputBytes, 0, bytesRead)
            fsOut.Write(finalOutputBytes, 0, finalOutputBytes.Length)

            ' Clear Base64Transform object.
            transf.Clear()
         End If

         Console.WriteLine()
         Console.WriteLine(My.Resources.CmdResources.SuccessEncodedMsg, srcFile, targetFile)
      Catch e As IOException
         Console.WriteLine(My.Resources.CmdResources.IOException)
         Exit Sub
      Catch e As SecurityException
         Console.WriteLine(My.Resources.CmdResources.SecurityException, srcFile)
         Exit Sub
      Catch e As UnauthorizedAccessException
         Console.WriteLine(My.Resources.CmdResources.AccessException, srcFile)
         Exit Sub
      Finally
         If sw IsNot Nothing Then sw.Close()
         If fs IsNot Nothing Then fs.Dispose() : fs.Close()
         If fsOut IsNot Nothing Then fsOut.Dispose() : fsOut.Close()
      End Try
   End Sub

   Private Sub DecodeFile(ByVal srcFile As String, ByVal targetFile As String)

      Dim reader As TextReader = Nothing
      Dim fs As FileStream = Nothing
      Dim fsIn As FileStream = Nothing
      Dim sourceSize As Long = New FileInfo(srcFile).Length

      Try
         fs = New FileStream(targetFile, FileMode.Create, FileAccess.Write)

         ' Read entire file in a single operation.
         If sourceSize <= CharsToRead Then
            reader = New StreamReader(srcFile, Encoding.ASCII)

            Dim encodedChars As String = reader.ReadToEnd()

            Try
               Dim bytes() As Byte = Convert.FromBase64String(encodedChars)
               fs.Write(bytes, 0, bytes.Length)
            Catch e As FormatException
               Console.WriteLine(e.Message)
            End Try

            Console.CursorLeft = 0
         Else
            ' Read file as a stream into a buffer.

            ' Instantiate a FromBase64Transform object.
            Dim transf As New FromBase64Transform()
            ' Arrays to hold input and output bytes.
            Dim inputBytes(transf.InputBlockSize - 1) As Byte
            Dim outputBytes(transf.OutputBlockSize - 1) As Byte
            Dim bytesRead As Integer = 0
            Dim bytesWritten As Integer = 0
            Dim totalBytesRead As Integer = 0

            fsIn = New FileStream(srcFile, FileMode.Open, FileAccess.Read)

            Do
               bytesRead = fsIn.Read(inputBytes, 0, inputBytes.Length)
               totalBytesRead += bytesRead
               bytesWritten = transf.TransformBlock(inputBytes, 0, bytesRead, outputBytes, 0)
               fs.Write(outputBytes, 0, bytesWritten)
            Loop While sourceSize - totalBytesRead > transf.InputBlockSize

            ' Transform the final block of data.
            bytesRead = fsIn.Read(inputBytes, 0, inputBytes.Length)
            Dim finalOutputBytes As Byte() = transf.TransformFinalBlock(inputBytes, 0, bytesRead)
            fs.Write(finalOutputBytes, 0, finalOutputBytes.Length)

            ' Clear Base64Transform object.
            transf.Clear()
         End If

         Console.WriteLine()
         Console.WriteLine(My.Resources.CmdResources.SuccessDecodedMsg, srcFile, targetFile)
      Catch e As IOException
         Console.WriteLine(My.Resources.CmdResources.IOException)
         Exit Sub

      Finally
         If reader IsNot Nothing Then reader.Close()
         If fs IsNot Nothing Then fs.Close() : fs.Dispose()
         If fsIn IsNot Nothing Then fsIn.Close() : fs.Dispose()
      End Try
   End Sub
End Module
